src/cocotb/share/lib/simulator/simulatormodule.cpp

```diff
@@ -838,6 +843,165 @@ static PyObject *log_level(PyObject *, PyObject *args) {
+        return -1;
+    }
+    if ((period_steps < 2) || (high_steps < 1) ||
+        (high_steps >= period_steps)) {
+        LOG_ERROR("Failed to start clock: invalid parameters");
+        return -1;
+    }
+
+    period = period_steps;
+    t_high = high_steps;
+
+    // Call toggle to start the clock, so set to the opposite value here.
+    clk_val = start_high ? 0 : 1;
+    return toggle();
+}
+
+int GpiClock::stop() {
+    if (!clk_toggle_cb_hdl) {
+        return -1;
+    }
+    gpi_deregister_callback(clk_toggle_cb_hdl);
+    clk_toggle_cb_hdl = nullptr;
+    return 0;
+}
+
+int GpiClock::toggle() {
+    clk_val = !clk_val;
+    gpi_set_signal_value_int(clk_signal, clk_val, GPI_DEPOSIT);
+
+    uint64_t to_next_edge = clk_val ? t_high : (period - t_high);
+
+    clk_toggle_cb_hdl =
+        gpi_register_timed_callback(&GpiClock::toggle_cb, this, to_next_edge);
+    if (!clk_toggle_cb_hdl) {
+        // LCOV_EXCL_START
+        LOG_ERROR("Clock will be stopped: failed to register toggle cb");
+        return -1;
+        // LCOV_EXCL_STOP
+    }
+
+    return 0;
+}
+
+int GpiClock::toggle_cb(void *gpi_clk) {
+    GpiClock *clk_obj = (GpiClock *)gpi_clk;
+    return clk_obj->toggle();
+}
+
+// Create a new clock object
+static PyObject *clock_create(PyObject *, PyObject *args) {
+    if (!gpi_has_registered_impl()) {
+        // LCOV_EXCL_START
+        PyErr_SetString(PyExc_RuntimeError, "No simulator available!");
+        return NULL;
+        // LCOV_EXCL_STOP
+    }
+
+    Py_ssize_t numargs = PyTuple_Size(args);
+
+    if (numargs != 1) {
+        PyErr_SetString(PyExc_TypeError,
+                        "Attempt to create Clock with wrong arguments!");
+        return NULL;
+    }
+
+    // Extract the clock signal sim object
+    PyObject *pSigHdl = PyTuple_GetItem(args, 0);
+    if (Py_TYPE(pSigHdl) != &gpi_hdl_Object<gpi_sim_hdl>::py_type) {
+        PyErr_SetString(PyExc_TypeError,
+                        "First argument must be a gpi_sim_hdl!");
+        return NULL;
+    }
+    gpi_sim_hdl sim_hdl = ((gpi_hdl_Object<gpi_sim_hdl> *)pSigHdl)->hdl;
+
+    GpiClock *gpi_clk = new GpiClock(sim_hdl);
+
+    if (gpi_clk) {
+        return gpi_hdl_New(gpi_clk);
+    } else {
+        // LCOV_EXCL_START
+        PyErr_SetString(PyExc_RuntimeError, "Failed to create clock!");
+        return NULL;
+        // LCOV_EXCL_STOP
+    }
+}
+
+static void clock_dealloc(PyObject *self) {
+    if (!gpi_has_registered_impl()) {
+        // LCOV_EXCL_START
+        PyErr_SetString(PyExc_RuntimeError, "No simulator available!");
+        return;
+        // LCOV_EXCL_STOP
+    }
+
+    if (Py_TYPE(self) != &gpi_hdl_Object<gpi_clk_hdl>::py_type) {
+        // LCOV_EXCL_START
+        PyErr_SetString(PyExc_TypeError, "Wrong type for clock_dealloc!");
+        return;
+        // LCOV_EXCL_STOP
+    }
+
+    GpiClock *gpi_clk = ((gpi_hdl_Object<gpi_clk_hdl> *)self)->hdl;
+
+    delete gpi_clk;
+
+    PyMem_Free(self);
```

Comment:
```suggestion
    Py_TYPE(self)->tp_free((PyObject *) self);
```
Or `PyObject_Del(self)` per https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_dealloc since this can't be subclassed.